### PR TITLE
ENH: Github Action to Spell Check Comments

### PR DESCRIPTION
**Pull request by dave3d**
This creates a new Github Action that spell checks the comments in header files when and code is checked in.  It uses the SimpleITK/CommentSpellCheck repo to do the spell checking of the Modules and Examples sub-directories.  Modules/ThirdParty is skipped.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [x] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: .github/workflows/itk_dict.txt [hunk 1]
```diff
@@ -0,0 +1,997 @@
...
+Shanhbag
+Shanmugam
+Shinin
+SinglephaseChanAndVeseSparseFieldLevelSetSegmentation
+SingleValuedVnlCostFunctionAdaptorv
+Smolikova
+Sobel
+Soille's
+Sonka
+Spall
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:193]: Could this be mis-spelling of `Spell`? Or is it a name?

File: .github/workflows/itk_dict.txt [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+Stiehl
+Syst
+Tajbakhsh
+Talos
+Tannenbaum
+Templateless
+Tfar
+Thiran
+Thuente
+Thuente's
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:205]: Given previous line, is this needed?

File: .github/workflows/itk_dict.txt [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+jvm
+k'th
+kHd
+kitware
+kwargs
+kwstyle
+kmeans
+kth
+labeler
+labeller
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:572]: Should this be removed? There is `labeler` on the previous line.

File: .github/workflows/itk_dict.txt [hunk 4]
```diff
@@ -0,0 +1,997 @@
...
+labelized
+lagrange
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:574]: Should this be capitalized?

File: .github/workflows/itk_dict.txt [hunk 5]
```diff
@@ -0,0 +1,997 @@
...
+orienter
+orthant
+orthantwise
+ortho
+orthogonalize
+orthonormality
+opencl
+openjpeg
+oper
+outJacobian's
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:709]: Maybe change instances of this and remove it from dictionary?

File: Modules/Segmentation/DeformableMesh/src/itkDeformableSimplexMesh3DGradientConstraintForceFiler.cxx [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/*=========================================================================
```
Comment (dzenanz) [inline Modules/Segmentation/DeformableMesh/src/itkDeformableSimplexMesh3DGradientConstraintForceFiler.cxx:1]: Was this file added intentionally? If so, it should probably be in a different commit, with its own message.

**Review by dzenanz — COMMENTED** [review]
It would be good to separate the spell check changes into an additional commit.

**Comment by dave3d** [comment]
Yeah I was trying to squash a bunch of commits into two separate commits, but I'm not very facile with got.

File: .github/workflows/itk_dict.txt [hunk 7]
```diff
@@ -0,0 +1,997 @@
...
+Calc
+Cardano's
+Chrisochoides
+Chu
+CiDxy
+Clatz
+Clunie
+Clunies
+CommandIterationUpdatev
+Comput
```
[the remark this answers is not in the document]
Comment (dave3d) [reply .github/workflows/itk_dict.txt:36]: That comes from here: https://github.com/InsightSoftwareConsortium/ITK/blob/82d7060408edd452b57b54bd96ee8e2fa037b474/Modules/Numerics/Optimizersv4/include/itkCommandIterationUpdatev4.h#L28C35-L28C35

File: .github/workflows/itk_dict.txt (hunk 1 above)
Comment (dave3d) [reply .github/workflows/itk_dict.txt:193]: Author name J.C. Spall:
https://github.com/InsightSoftwareConsortium/ITK/blob/82d7060408edd452b57b54bd96ee8e2fa037b474/Modules/Numerics/Optimizers/include/itkSPSAOptimizer.h#L57

File: .github/workflows/itk_dict.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (dzenanz) [reply .github/workflows/itk_dict.txt:36]: I was talking about `Comput`, not `CommandIterationUpdatev`.

File: .github/workflows/itk_dict.txt (hunk 5 above)
Comment (dave3d) [reply .github/workflows/itk_dict.txt:709]: outJacobian is a variable name that is referred to in the comments.  I don't want to change that.
https://github.com/InsightSoftwareConsortium/ITK/blob/82d7060408edd452b57b54bd96ee8e2fa037b474/Modules/Core/Transform/include/itkCompositeTransform.h#L400

File: .github/workflows/itk_dict.txt (hunk 3 above)
Comment (dave3d) [reply .github/workflows/itk_dict.txt:572]: It seems that 'labelled' is acceptable British spelling, and so I guess 'labeller' ought to be ok too.  The code has a lot of uses of 'labelled' in it, even in method and variable names.

File: Modules/Segmentation/DeformableMesh/src/itkDeformableSimplexMesh3DGradientConstraintForceFiler.cxx (hunk 6 above)
Comment (dave3d) [reply Modules/Segmentation/DeformableMesh/src/itkDeformableSimplexMesh3DGradientConstraintForceFiler.cxx:1]: Oops.  That's a mistake.

File: .github/workflows/itk_dict.txt [hunk 8]
```diff
@@ -0,0 +1,994 @@
...
+jacobians
+jcd
+jcg
+jh
+jj
+jsi
+jvm
+k'th
+kHd
+kitware
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:565]: This should be capitalized.

File: .github/workflows/itk_dict.txt [hunk 9]
```diff
@@ -0,0 +1,994 @@
...
+quantized
+qto
+radiographs
+radiological
+randgen
+rasterization
+rasterized
+rasterize
+rasterizable
+rayleigh
```
Comment (dzenanz) [inline .github/workflows/itk_dict.txt:765]: Should this be capitalized?

**Review by dzenanz — COMMENTED** [review]
Thanks for the improvements. I just spotted two more 😄

**Comment by dzenanz** [comment]
There is also [this](https://github.com/InsightSoftwareConsortium/ITK/actions/runs/6338791879/job/17216608327?pr=4229), caught by the action itself 😄

**Comment by dave3d** [comment]
> There is also [this](https://github.com/InsightSoftwareConsortium/ITK/actions/runs/6338791879/job/17216608327?pr=4229), caught by the action itself 😄

Yeah, after I took is out of the dictionary. Oh well.

**Comment by dzenanz** [comment]
Replace it by the full title, `IEEE Transactions on Evolutionary Computation`?

**Comment by dave3d** [comment]
https://github.com/InsightSoftwareConsortium/ITK/blob/75b99671a89a9ef6ba1ad8fda62ba3843ad61b54/Examples/RegistrationITKv4/ImageRegistration8.cxx#L396
I put 'kitware' back as lowercase because of the above line.

**Comment by dzenanz** [comment]
Domain names are not case sensitive, we could use uppercase there. Also, that link is now broken, and should be replaced by this link: `https://data.Kitware.com/#collection/57b5c9e58d777f126827f5a1/folder/57b769c08d777f10f269af3a`.

**Review by dzenanz — APPROVED** [review]

File: .github/workflows/spell-check-comments.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    - name: Set up Python 3.9
+      uses: actions/setup-python@v4
+      with:
+        python-version: 3.9
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install git+https://github.com/SimpleITK/CommentSpellCheck@main
```
Comment (blowekamp) [inline .github/workflows/spell-check-comments.yml:23]: We should not use an untagged version. Also installing over git is slow. Use the a package from pypi.

**Review by blowekamp — COMMENTED** [review]
I think the spelling changes should be made in another PR, and merge quickly. The CI need a little more work.

File: .github/workflows/spell-check-comments.yml (hunk 10 above)
Comment (dave3d) [reply .github/workflows/spell-check-comments.yml:23]: OK, I haven't yet gotten it into pypi.  That is the next thing to do.

**Review by thewtex — COMMENTED** [review]
LGTM, thanks for contributing this.

I agree with @blowekamp that we should use a PyPI released version before merging.

**Comment by dzenanz** [comment]
Rebase now on master, as #4252 is merged? Also possibly convert to draft until PyPI package is available?

**Comment by dave3d** [comment]
I think, maybe, possibly, hopefully, that this PR might finally be good to go.

**Review by dzenanz — APPROVED** [review]

**Comment by dzenanz** [comment]
```log
Submit files
   SubmitURL: https://open.cdash.org/submit.php?project=Insight
   Upload file: D:/a/1/s-build/Testing/20231023-1531/Update.xml to https://open.cdash.org/submit.php?project=Insight&FileName=Azure.fv-az676-359___Windows_NT-Build10533-PR4229-SpellcheckAction___20231023-1531-Experimental___XML___Update.xml&build=Windows_NT-Build10533-PR4229-SpellcheckAction&site=Azure.fv-az676-359&stamp=20231023-1531-Experimental&MD5=9331b22b8e7d9814849a2bc33c565aba Size: 659
   Error when uploading file: D:/a/1/s-build/Testing/20231023-1531/Update.xml
   Error message was: server certificate verification failed. CAfile: none CRLfile: none
   Problems when submitting via HTTP
##[error]Cmd.exe exited with code '-1'.
Finishing: Build and test
```
Does anyone know what is wrong, and how to fix it?
